### ISSUE TITLE
Finalize explainability output

### ISSUE THREAD

**Issue by ShiriMoran** [np-guard/vpc-network-config-analyzer#357]
E.g.:
```
Connectivity explanation between vsi1-ky[10.240.10.4] and vsi2-ky[10.240.20.4]
No connection between vsi1-ky[10.240.10.4] and vsi2-ky[10.240.20.4]; connection blocked by egress
Egress:
~~~~~~~~~
network ACL acl1-ky enables egress
relevant rules from network ACL acl1-ky:
        index: 0, direction: outbound , src: 0.0.0.0/0 , dst: 0.0.0.0/0, conn: all, action: allow
security group sg1-ky disables egress
Ingress:
~~~~~~~~~
network ACL acl2-ky enables ingress
relevant rules from network ACL acl2-ky:
        index: 1, direction: inbound , src: 0.0.0.0/0 , dst: 0.0.0.0/0, conn: all, action: allow
security group sg2-ky enables ingress
relevant rules from security group sg2-ky:
        index: 4, direction: inbound, protocol: all, cidr: 10.240.10.4/32

```

Note that relevant rules can be for nacl both disabling and enabling rules; specifically, it can be a combination of allow and deny rules resulting in allowing or in disabling the direction

In 2nd thought - this extra information is relevant only if the direction is blocking. 
There are 3 options here:
1. Print the addition "xxxx enables ingress/egress" always, for any relevant filter
2. Print it only if the **entire connection** is blocked
3. Print it for **ingress/egress** only if the direction is blocking 
The example above will look the same in the 1st two options; in the 3rd option the lines "network ACL acl2-ky enables ingress" and "security group sg2-ky enables ingress" will not be there

I vote for 2 
@adisos @zivnevo

**Comment by zivnevo** [comment]
As a user I would like to get the following pieces of information **in this order**:
1. Is the connection blocked at egress, ingress or both
2. Is it a filter or is it a (missing/misconfigured) router that prevents the connection (or maybe both)?
3. If it's filter, which filter (SG/NACL, or both)?  If it's a router - which router is missing / misconfigured?
4. If it's a filter, is it because there is no allow rule or because there is a deny rule (include this info even if the filter is SG).
5. Filters and routers that do allow the connection + relevant details (e.g., allow rules)

Adding @kyorav to this discussion

**Comment by ShiriMoran** [comment]
Comments:

> 
>     2. Is it a filter or is it a (missing/misconfigured) router that prevents the connection (or maybe both)?
> 
"both" is not well defined. Namely, if a lack of router prevents the connection then I do not know which filters are relevant.
Specifically, I do not know whether ACL is relevant if the connection is to the outside world and there is no router.
Thus, if there is no router then this is all we report (at the moment). I think adding a "what if" for each potential router will be an over kill in this stage. I can add the set of relevant rules with no context - 3 and 4 - I think for the first release better just keep as is, and see what feedback we get
>     3. If it's filter, which filter (SG/NACL, or both)?  If it's a router - which router is missing / misconfigured?
Which router you mean "fip" or "fip and pgw"?  (as in the demo I had)
What do you mean be misconfigured?

> 
>     4. If it's a filter, is it because there is no allow rule or because there is a deny rule (include this info even if the filter is SG).
> 
>     5. Filters and routers that do allow the connection + relevant details (e.g., allow rules)
What about rules that deny the connection? I think we should present these as well
 

As to the order: I'm OK with having 1-3 in this order.
However, once we get to specific tables - it might be confusing to have 4 for and then 5 for all tables.  Perhaps it will be better to have 4 and 5 should be per table. Also given that 5 might include deny rules. 
I think that after 1-3 we should present each relevant table with its details:
4. allows/denies relevant connection, if denies - due to lack of allow rules or due to deny rules
5. rules

Also, I would add the following *before* getting into filter tables details:
3'. If there is an enabling router - which is it (type + name)
I would not mix the router (which is single and simple) with the tables (which could be multiple and each of them has many details attached to it)

**Comment by kyorav** [comment]
I agree with Ziv's ordering. A couple of comments:
- For this output to be useful we should try to present as little information as needed and in the most concise way.
- It seems the report is for a directed connection, in which case it would be best if it was clearly worded that way (e.g. "No connection *from* vsi1 *to* vsi2...")
- I am not sure about item 5, it will add a lot of distraction. I think that if the connection is not possible we should only explain why it isn't possible and for a connection that is possible we should only explain why it is. 

> 2. Is it a filter or is it a (missing/misconfigured) router that prevents the connection (or maybe both)?

If there is no route from src to destination then this is the most important piece of information. And since without a route to destination we can't always determine which filters are applied it makes sense not to mention them in this scenario.

> 3. If it's filter, which filter (SG/NACL, or both)?  If it's a router - which router is missing / misconfigured?

If the src/dst is a public IP address then you just print "no FIP or Public Gateway". When the analysis supports transit gateways there will be the possibility that the src and dst are in different VPCs that aren't connected to the same gateway. And when we have routing tables in the analysis we will print the name of the table that is applied and the relevant route in the table.

> What about rules that deny the connection (deny rules can be part of a denied connection and of an allow connection; allow rules can not be part of an denied connection)?

I don't know what "deny rule" refers to in this comment, but in the spirit of minimum information I suggest the following:
- for an allowed connection we print the names of all ACLs and SGs that are evaluated on the way (in the order of evaluation). For each ACL we mention the name of the rule that catches this connection according to the prioritized evaluation, and for each SG the name of the rule that allows this connection 
- for a connection that is blocked we print all the ACLs and SGs that are blocking (according to the order of evaluation), and for each such ACL the name of the rule that blocks (or "blocked by default"). Maybe, if we add a "verbose" mode, we print *all* the ACLs and SGs on the route and write "allowing" next to those that allow.

**Comment by ShiriMoran** [comment]
comments:
 @zivnevo I see Karen agrees with me that in case there is no router to enable the required connection there is no point in referencing the filters which are not well defined. 
@kyorav: |
1. deny rules are ACL rules that explicitly denies certain traffic
2. Printing the names of the rules does not sound like a good idea, mainly because in SG there are no names to the rules [see here](https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/swagger-ui.html?version=today#/Security%20groups/create_security_group)
>     * for an allowed connection we print the names of all ACLs and SGs that are evaluated on the way (in the order of evaluation). For each ACL we mention the name of the rule that catches this connection according to the prioritized evaluation, and for each SG the name of the rule that allows this connection
4. Not do you mean by "order of evaluation"?
5. There could be multiple rules in each filter that enables or disables a connection. In fact, an enabled connection may have "deny" and "allow" ACL rules which are relevant for it
6. Verbosity - we can either add a command line parm or just use the debug output format we already have

Anyways, once we have all the relevant or could be relevant information in handy struct, fine tuning the output should not be big deal. In fact it could be easier once we have a demo (which should be available within a few days).

**Comment by kyorav** [comment]
3. Filters are evaluated in a particular order on the route from src to dst. In the case where the src and dst are NIFs in different subnets this would be: (1) the SGs attached to the src NIF (2) the outbound rules in the ACL attached to the src NIF's subnet (3) the inbound rules in the ACL attached to the dst NIF's subnet (4) the SGs attached to the dst NIF. As a user trying to understand the symptoms I am seeing it will help me if the information is presented in the actual order of execution. Note that when we support routing tables they will fall somewhere in the middle, I believe 2.5. 

4. In a network ACL only one rule determines the fate of this connection. Rules are evaluated in order, the first one that matches calls the shots and the rest are ignored. I think that in the default output we should only print the one that matches. Perhaps in a "verbose" mode we might print all those that may become relevant if the one that matched was removed. With SGs it's simpler, all those that match are relevant for an enabled connection. 
However, if the connection is blocked by an ACL while the SG allows it, I think we shouldn't print the SG information because it does not help explain why the connection is blocked.

> Anyways, once we have all the relevant or could be relevant information in handy struct, fine tuning the output should not be big deal. In fact it could be easier once we have a demo (which should be available within a few days).

👍

**Comment by ShiriMoran** [comment]
@kyorav 
3. Thanks 
4. If a connection refers to a single protocol and a single `src` and `dst` port then indeed a single ACL rule determines its fate. This is not the case otherwise. We support querying for only `src` and `dst` in which we consider any possible connection in between, or for a specific protocol and any range of `src` and `dst` ports (one may not specify any or all of the ports parms, in which it is assumed to be the min/max value). We may decide to limit the 1st release to a single `src` and `dst` `port,` though I don't see why.

Anyways, I see no point in printing ACL rules that have no real effect - it might only confuse the user. If we do print it, we should mention they have no effect in the current setting.

**Comment by ShiriMoran** [comment]
solved by #374
